### PR TITLE
upgrade jest-editor-support and multiroot status

### PR DESCRIPTION
**Pull request by connectdotz**
This PR mainly to 
1. integrate with the latest `'jest-editor-support'`, which contains many fixes needed by multiroot (#442), template-literal(#247), windows useWsl (#393)
1. address multiroot status(#379) 
1. paying some technical debt...

changes by category:

## multiroot
- StatusBar reports both workspace summary and active folder jest status. Added new summary output channel to give full details of each folder status.
- integrate with the latest `jest-editor-support' to pick up concurrency and others fixes from [jest-editor-support/#9](https://github.com/jest-community/jest-editor-support/pull/9)

## misc improvement
- support template-literal-string test names (#247)
  - when failed to match template-literal-string test names, it can now fall back to matching by locations reported by jest.

## technical debt (internal development improvement)
- adopts ts-jest for better sourceMap support and more 
- adopt the latest 'tslint' rules for best practice coding style.
- migrate to typescript 3.x and jest 24.7.x
- some tests clean up to consolidate mocks and others.

---
Note:
- This PR has a dependency on [jest-editor-support/#9](https://github.com/jest-community/jest-editor-support/pull/9), the build will fail until a new release of 'jest-editor-support' is cut.
- there are lots of files touched due to typescript/lint changes, most to conform to typescript 3.x and lint rules.

### CODE REVIEW

**Comment by connectdotz** [comment]
Regarding the multiroot-status, I ended up displaying 2 separate statusBarItem - active folder and workspace summary - each command opens up an output channel for more details. 

- The summary status is pretty much @escaton original implement, I just added a new output channel to display all folder's status when clicked. 
- The active folder status is more dynamic, mainly depends on the active editor. The status will open the corresponding output channel when clicked. Unlike single-root, this status will not display the long details strings.

Here is a snapshot since you guys won't be able to build this PR yet (due to jest-editor-support). Please don't hesitate to suggest how we can further improve it...

multiroot-status

(single-root projects status will appear identical to the production version)

**Comment by connectdotz** [comment]
This PR is rather large, maybe we could have multiple people reviewing it?

@escaton, do you mind reviewing the mutiroot related files, such as extensionManager, StatusBar, extension, JestExt? 

@orta or @seanpoulter or @stephtr, I know it is a pain to go through so many files, but many exciting new features are riding on these PRs, when you get a chance, do you mind reviewing both [jest-editor-support/#9](https://github.com/jest-community/jest-editor-support/pull/9) and this one. Thanks.

File: src/StatusBar.ts [hunk 1]
```diff
@@ -1,26 +1,62 @@
...
 
 type Status = 'running' | 'failed' | 'success' | 'stopped' | 'initial'
-type QueueItem = {
+interface QueueItem {
   source: string
   status: Status
   details: string | undefined
 }
 
+interface StatusBarSpinner {
```
Comment (escaton) [inline src/StatusBar.ts]: Not the best name for such composite entity. I'd prefer to have two separate spinners, like `StatusItem` below

File: src/StatusBar.ts [hunk 2]
```diff
@@ -1,26 +1,62 @@
...
+    this.summaryStatusItem.command = showSummaryOutput
+    this.activeStatusItem.command = showAciiveOutput
+
+    return [
+      vscode.commands.registerCommand(showSummaryOutput, () => {
+        if (this.workspaceOutput) {
+          this.workspaceOutput.show()
+        }
+      }),
+      vscode.commands.registerCommand(showAciiveOutput, () => {
```
Comment (escaton) [inline src/StatusBar.ts]: typo

File: src/StatusBar.ts [hunk 3]
```diff
@@ -41,32 +77,156 @@ export class StatusBar {
...
+      }
+      return
+    }
+
+    // find the active item from the queue
+    const queueItem = this.activeFolder
+      ? this.queue.find(_item => _item.source === this.activeFolder)
+      : this.queue.length === 1
+      ? this.queue[0]
+      : undefined
```
Comment (escaton) [inline src/StatusBar.ts]: If this is not an explicit code style, i believe some indentation may be appropriate here :)

File: src/StatusBar.ts [hunk 4]
```diff
@@ -41,32 +77,156 @@ export class StatusBar {
...
+
+    const messages = this.queue.map(item => {
+      const details = item.details ? `: ${item.details}` : ''
+      return `${item.source}: ${item.status} ${details}`
+    })
+    this.workspaceOutput.append(messages.join('\n'))
   }
+
+  private isMultiroot() {
+    return this.queue.length > 1
```
Comment (escaton) [inline src/StatusBar.ts]: this is implementation detail, it needs to be more explicit

File: src/StatusBar.ts [hunk 5]
```diff
@@ -41,32 +77,156 @@ export class StatusBar {
...
     this.queue = this.queue.filter(x => x.source !== source)
-    this.queue.unshift({
+    const item: QueueItem = {
       source,
       status,
       details,
-    })
-    this.updateStatus()
+    }
+    this.queue.unshift(item)
```
Comment (escaton) [inline src/StatusBar.ts]: i just realize, that there is a memory leakage here. We should filter out all previous items with same source. Would you mind to add it?

File: src/StatusBar.ts (hunk 1 above)
Comment (connectdotz) [reply src/StatusBar.ts]: I agree. Actually, I think there are simply too many 'get x by statusType' kind of logic throughout the code, time to do some refactoring to consolidate those into a self-contained block...

File: src/StatusBar.ts (hunk 2 above)
Comment (connectdotz) [reply src/StatusBar.ts]: good catch

File: src/StatusBar.ts (hunk 3 above)
Comment (connectdotz) [reply src/StatusBar.ts]: even with indentation, I agree it could still be hard to read, sigh, breaking it up...

File: src/StatusBar.ts (hunk 4 above)
Comment (connectdotz) [reply src/StatusBar.ts]: not sure what you mean here... I wanted to check if there are multiple sources to decide if to show the summary status. It might make more sense after switching queue to Map (see below)

File: src/StatusBar.ts (hunk 5 above)
Comment (connectdotz) [reply src/StatusBar.ts]: I think we can be more efficient with a Map, not array/queue here. Since we really just need 1 status per source and the sequence is not important since you used the status priority as the determining factor... take a look of the updated StatusBar and see if it makes sense...

File: tests/StatusBar.test.ts [hunk 6]
```diff
@@ -29,10 +33,31 @@ describe('StatusBar', () => {
...
+      registerCommand.mock.calls.forEach(c => {
+        // tslint:disable no-bitwise
+        if (c[0].includes('show-summary-output')) {
+          found |= 0x1
+        }
+        if (c[0].includes('show-active-output')) {
+          found |= 0x2
+        }
+      })
+      expect(found).toEqual(3)
```
Comment (stephtr) [inline tests/StatusBar.test.ts]: It is more performant, but is it also easier to read than something like:
```
const calls = registerCommand.mock.calls
expect(calls.some(c => c[0].includes('show-summary-output'))).toBe(true)
expect(calls.some(c => c[0].includes('show-active-output'))).toBe(true)
```
?

File: tests/StatusBar.test.ts [hunk 7]
```diff
@@ -42,26 +67,12 @@ describe('StatusBar', () => {
       const helpers = statusBar.bind(source)
       ;['initial', 'running', 'success', 'failed', 'stopped'].forEach(status => {
         helpers[status]()
-        expect((statusBar as any).queue).toContainEqual({ source, status })
+        expect(updateStatusSpy).toHaveBeenCalledWith({ source, status })
```
Comment (stephtr) [inline tests/StatusBar.test.ts]: Are you sure that this is working? `{ source, status }` creates a new object and I thought that `toHaveBeenCalledWith` only checks identity of the reference (in contrast to `toContainEqual`)

File: src/StatusBar.ts [hunk 8]
```diff
@@ -1,72 +1,254 @@
...
+  clearSpinner(): void
+  startSpinner(callback: () => void): void
+}
+
+const createStatusBarItem = (type: StatusType, priority: number): SpinnableStatusBarItem => {
+  let spinner: NodeJS.Timer | undefined
+  const item = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, priority)
+  const clearSpinner = () => {
+    if (spinner) {
+      clearInterval(spinner)
```
[the remark this answers is not in the document]
Comment (connectdotz) [reply src/StatusBar.ts]: in statusBar `render()` , it did call the `clearSpinner()` so basically each `startSpinner` only call back for 1 interval unless it is started again within `render()`. Does that address your conern?

File: src/TestResults/TestResultProvider.ts [hunk 9]
```diff
@@ -31,26 +42,65 @@ export class TestResultProvider {
   }
 
   getResults(filePath: string): TestResult[] {
+    const maybeTemplateLiteral = (s: string) => s.indexOf('$') > -1
```
[the remark this answers is not in the document]
Comment (connectdotz) [reply src/TestResults/TestResultProvider.ts]: that's a good suggestion, done

File: src/TestResults/TestResultProvider.ts [hunk 10]
```diff
@@ -31,26 +42,65 @@ export class TestResultProvider {
...
+              err = 'failed to match test result, might be caused by template-literal test name?'
+            }
+          }
+          break
+        default:
+          // multiple matches, select according to the following criteria
+          assertion = findAssertionByLocation(test, match)
+          if (!assertion) {
+            // can't find the match, it could due to sourceMap related issue, let's try our best to locate one then
+            assertion = match.find(a => a.status !== TestReconciliationState.KnownFail) || match[0]
```
[the remark this answers is not in the document]
Comment (connectdotz) [reply src/TestResults/TestResultProvider.ts]: hmm... good question, I don't know the answer either... looks like it was added in this [commit](https://github.com/jest-community/vscode-jest/commit/4a5d225a31930b091e1fa0a0be85414bf66e2b82#diff-785a64be7997ebe2d7a9fdf2b7231468), maybe @ThomasRooney can shed some light?

File: src/StatusBar.ts [hunk 11]
```diff
@@ -1,72 +1,254 @@
...
+    statusBarItem.clearSpinner()
+
+    const message = this.getMessageByStatus(request.status)
+    if (request.status === 'running') {
+      statusBarItem.startSpinner(() => this.render(request, statusBarItem))
+    }
+
+    switch (statusBarItem.type) {
+      case StatusType.active:
+        const details = !this.needsSummaryStatus() && request.details ? request.details : ''
```
[the remark this answers is not in the document]
Comment (connectdotz) [reply src/StatusBar.ts]: I didn't see any lint rules/best-practice regarding the condition in ternary expression, however, did find the [multiline-ternary](https://eslint.org/docs/rules/multiline-ternary) in eslint, which defaults to separate lines. I could change it accordingly, and it will probably make the condition more readable.

File: src/extensionManager.ts [hunk 12]
```diff
@@ -6,13 +6,15 @@ import { JestExt } from './JestExt'
 import { DebugCodeLensProvider, TestState } from './DebugCodeLens'
 import { DebugConfigurationProvider } from './DebugConfigurationProvider'
 import { IPluginResourceSettings, IPluginWindowSettings } from './Settings'
+import { statusBar } from './StatusBar'
 
 export class ExtensionManager {
+  debugCodeLensProvider: DebugCodeLensProvider
+  debugConfigurationProvider: DebugConfigurationProvider
+
   private extByWorkspace: { [key: string]: JestExt } = {}
```
[the remark this answers is not in the document]
Comment (connectdotz) [reply src/extensionManager.ts]: from the assignment code in `register()`:
```
this.extByWorkspace[workspaceFolder.name] = new JestExt(
```
it should always be `JestExt`... not all workspace will be in the "extByWorkspace" dictionary, i.e. not all worksapces will be registered.

File: src/extensionManager.ts [hunk 13]
```diff
@@ -79,15 +82,18 @@ export class ExtensionManager {
...
     return (
       // start if enabledWorkspaceFolders contains current workspace
       enabledWorkspaceFolders.includes(workspaceFolderName) ||
-      // dont start if disabledWorkspaceFolders doesn't contain current workspace
+      // don't start if disabledWorkspaceFolders doesn't contain current workspace
       !disabledWorkspaceFolders.includes(workspaceFolderName)
     )
   }
-  getByName(workspaceFolderName: string) {
+  public getByName = (workspaceFolderName: string) => {
```
[the remark this answers is not in the document]
Comment (connectdotz) [reply src/extensionManager.ts]: arrow-function can capture "this" so it is easier to pass alone as a function-parameter, such as in the `extension.activate()`:
```
context.subscriptions.push(...statusBar.register(extensionManager.getByName),
```
instead of:
```
context.subscriptions.push(...statusBar.register((workspaceFolderName: string) => extensionManager.getByName),
```

File: tests/StatusBar.test.ts (hunk 6 above)
Comment (connectdotz) [reply tests/StatusBar.test.ts]: I like that, especially in tests I agree we should strive for clarity than performance... done

File: tests/StatusBar.test.ts (hunk 7 above)
Comment (connectdotz) [reply tests/StatusBar.test.ts]: as far as jest 24.7.0 goes, yes it worked, otherwise, the test will fail 😏

File: src/StatusBar.ts (hunk 5 above)
Comment (connectdotz) [reply src/StatusBar.ts]: @escaton did you get the chance to look at the updated file? does that address your request?

**Comment by connectdotz** [comment]
ok, we have released the `jest-editor-support` and now this PR is finally able to be compiled and passed ci tests. Thanks for @stephtr and @escaton to review it, please take a look at the latest commit to see if you have any concern that hasn't been addressed yet.

File: src/StatusBar.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (stephtr) [reply src/StatusBar.ts]: My idea was to look at the statusBarItem as a separate object. If for example the implementation of `StatusBar` would some when change and by chance call `startSpinner` twice, the item would spin indefinitely. Therefore I would check within `startSpinner` that `spinner` is `undefined`. But that's just my personal opinion and to be honest completely unimportant.

File: src/StatusBar.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (stephtr) [reply src/StatusBar.ts]: `const details = (!this.needsSummaryStatus() && request.details) || ''` probably wouldn't be obvious enough?
Otherwise I'm also fine with leaving it as it is.

File: src/extensionManager.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (stephtr) [reply src/extensionManager.ts]: I was thinking about `extByWorkspace['not_existing_workspace'], but that was my mistake.

File: src/extensionManager.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (stephtr) [reply src/extensionManager.ts]: Ah ok, I didn't notice that.
(Somehow I still slightly prefer your second option or `.register(extensionManager.getByName.bind(extensionManager))` over changing an object method to an error function just that is easier to call it, but that was just my personal stubborn opinion.)

File: src/Coverage/CoverageCodeLensProvider.ts [hunk 14]
```diff
@@ -2,15 +2,18 @@ import * as vscode from 'vscode'
...
 
-  constructor(getJestExt: (uri: vscode.Uri) => JestExt) {
+  constructor(getJestExt: GetJestExt) {
     this.getJestExt = getJestExt
   }
 
   public provideCodeLenses(document: vscode.TextDocument, _token: vscode.CancellationToken) {
-    const coverage = this.getJestExt(document.uri).coverageMapProvider.getFileCoverage(document.fileName)
+    const ext = this.getJestExt(document.uri)
+    const coverage = ext ? ext.coverageMapProvider.getFileCoverage(document.fileName) : undefined
```
Comment (stephtr) [inline src/Coverage/CoverageCodeLensProvider.ts]: Since we return if `coverage` is falsy, what about `const coverage = ext && ext.coverageMapProvider.getFileCoverage(document.fileName)` instead?

File: src/DebugCodeLens/DebugCodeLensProvider.ts [hunk 15]
```diff
@@ -7,12 +7,13 @@ import { TestReconciliationState } from '../TestResults'
 import { TestState, TestStateByTestReconciliationState } from './TestState'
 import { JestExt } from '../JestExt'
 
+type GetJestExt = (uri: vscode.Uri) => JestExt | undefined
```
Comment (stephtr) [inline src/DebugCodeLens/DebugCodeLensProvider.ts]: Since it's the second time defined that way, would it be worth the effort to outsource it (for example to `vscode-jest/JestExt.ts`)?

File: src/Coverage/CoverageCodeLensProvider.ts (hunk 14 above)
Comment (connectdotz) [reply src/Coverage/CoverageCodeLensProvider.ts]: great idea!

File: src/DebugCodeLens/DebugCodeLensProvider.ts (hunk 15 above)
Comment (connectdotz) [reply src/DebugCodeLens/DebugCodeLensProvider.ts]: we could,  but it is not a method defined in `JextExt` but from `extensionManager`, so I move the type alias there instead. 

The real concern here is the fact that typescript didn't catch the original `undefined` type error and caused run time crash so I refactor the code and introduced an explicit typedef, in my opinion, a bit unnecessary if tsc does its job right...

File: src/StatusBar.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (connectdotz) [reply src/StatusBar.ts]: i see what you mean, and I think you are right, it is definitely more defensible for future development, in case spinner did get misused... added `clearSpinner` at the beginning of each `startSpinner` should do the trick.

**Comment by connectdotz** [comment]
sorry, was busy in the last few days, I need to re-merge with the latest beta changes and then we should be good to go...

**Comment by stephtr** [comment]
To me it seems like [your merge](https://github.com/jest-community/vscode-jest/pull/447/commits/74fcb6605fab50ab615356b78d9e1848663ccf42) went wrong, undoing previous changes.

File: src/TestResults/TestResultProvider.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (connectdotz) [reply src/TestResults/TestResultProvider.ts]: I think I figure this one out, I think it was assuming the "failed" test will have the line number matched already, so it is only focused on matching the non-failed ones, but of course, this assumption might not always be true since any location matching could fail due to transpile/sourcemap... 

while adding more tests to this, I discover an edge case for duplicate test names. Also didn't feel comfortable about template-literal use case, which right now will fall back to location matching, kind of dangerous as mentioned above. Added partial name match to increase correctness, when in doubt (such as multiple partial matched names found), we will just mark it as 'unknown' and log the info in console (in debug mode).

@stephtr if you get a chance, please take a look to make sure I didn't miss anything obvious, appreciate it.

**Comment by connectdotz** [comment]
ci failed due to test coverage rounding, anyway, I decided to beef up the tests coverage (we are +2.2% now 😄).  Plus a bonus: discovered a new bug in TestResultProvider, good karma 🙏 ...
